sql/src/main/java/io/crate/jobs/JobExecutionContext.java

```diff
@@ -151,13 +156,29 @@ public long keepAlive() {
         return this.keepAlive;
     }
 
+    public void kill() {
+        if (!closed.getAndSet(true)) {
+            if (activeSubContexts.get() == 0) {
+                callContextCallback();
+            } else {
+                for (ExecutionSubContext executionSubContext : subContexts.values()) {
+                    // kill will trigger the ContextCallback onClose too
+                    // so it is not necessary to remove the executionSubContext from the map here as it will be done in the callback
+                    executionSubContext.kill();
+                }
+            }
+        }
+    }
+
     public void close() {
-        LOGGER.trace("close called on JobExecutionContext {}", jobId);
-        if (activeSubContexts.get() == 0) {
-            callContextCallback();
-        } else {
-            for (ExecutionSubContext executionSubContext : subContexts.values()) {
-                executionSubContext.close();
+        if (!closed.getAndSet(true)) {
+            LOGGER.trace("close called on JobExecutionContext {}", jobId);
+            if (activeSubContexts.get() == 0) {
+                callContextCallback();
+            } else {
+                for (ExecutionSubContext executionSubContext : subContexts.values()) {
+                    executionSubContext.close();
+                }
```

Comment:
afaik all method which are changing the activeSubContexts must be aware of the closed boolean